### PR TITLE
feat: Add useMultipleSelect hooks, transfer supports multiple select by holding down the shift key.

### PR DESCRIPTION
**Pull request by zbw-zbw**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close https://github.com/ant-design/ant-design/issues/44669

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     Add useMultipleSelect hooks, transfer Supports multiple selection by holding down the shift key      |
| 🇨🇳 Chinese |     新增 useMultipleSelect hooks，transfer 支持按住 shift 键进行多选      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 14ccec0</samp>

This pull request refactors the selection logic for the `Table` and `Transfer` components using a custom hook `useMultipleSelect`, which handles the multiple selection with the shift key and the previous selected index. It also improves the code style, test coverage, and functionality of these components and their demos, and updates the documentation to reflect the new event parameter for the `onItemSelect` prop.

### <samp>🤖 Generated by Copilot at 14ccec0</samp>

*  Implement multiple selection logic for table and transfer components using useMultipleSelect hook ([link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25R13),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25R113-R115),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L535-R542),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdR5),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdR173-R175),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL278-R353))
*  Import useMultipleSelect hook from _util/hooks folder and useRef hook from React ([link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-4047d2a1719d0b72c656b76c12dc0e4787dba87a91f877ee49205e3d79c8f8c6R1-R59),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L2-R2))
*  Add event parameter to onItemSelect prop of table, transfer, and list components, and pass it to multipleSelect function when shift key is pressed ([link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L622-R588),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296R155-R221),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-edca56351272e05c831757bdcbfe5641e3d7615af8bfaa391b4ad9ef8b9eb49cL67-R69),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-a3fcfc2351816508d8b94df43d29a886ee240f42726572e193634aa2ce09ce75L67-R69),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-5d2a9fbb1586f5058bfd77d20e281f359ce068ed333ccc4375ee0ef1961342cbL59-R64),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-c165dbeb24339c042fcc975a9b661af8947ea942457cf8935f29817ac86033a9L71-R78),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdR195-R202),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdR274),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-f25eb82686b163c8fb19e3d453b3c168ce5311046d76e7b3a1192173b676ea97L74-R81),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415L55-R56),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL76-R77),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-0580116b55eb3d023efc52e18a28e0f852a621b7c4a6bce12ae1ddb163a28723L16-R16),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-0580116b55eb3d023efc52e18a28e0f852a621b7c4a6bce12ae1ddb163a28723L70-R70))
*  Update or reset previous selected index for each direction when onSelect, onClear, moveToLeft, moveToRight, or onItemSelectAll props are called ([link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L353-R354),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L409-R410),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL234-R251))
*  Remove lastSelectedKey state variable and setLastSelectedKey function, and remove lastSelectedKey from useMemo dependency array ([link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L220-L222),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L727))
*  Add and remove test cases for table row selection feature ([link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-c9dafe6d2a68bda33046f3f0bf89abcc077581df1c76fcb865501a7d6ec3c005L296-R339),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-c9dafe6d2a68bda33046f3f0bf89abcc077581df1c76fcb865501a7d6ec3c005L343-L352))
*  Reorder import statements to follow convention of importing React first, then other modules, then relative paths ([link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-c9dafe6d2a68bda33046f3f0bf89abcc077581df1c76fcb865501a7d6ec3c005L2-R6),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L2-R2),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296L1-R4),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-edca56351272e05c831757bdcbfe5641e3d7615af8bfaa391b4ad9ef8b9eb49cL1-R5),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-a3fcfc2351816508d8b94df43d29a886ee240f42726572e193634aa2ce09ce75L2-R5),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415L1-R5),[link](https://github.com/ant-design/ant-design/pull/45285/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL1-R12))

### CODE REVIEW

File: components/transfer/index.en-US.md [hunk 1]
```diff
@@ -68,14 +68,14 @@ Common props ref：[Common props](/docs/react/common-props)
...
-| onItemSelect    | Select item             | (key: string, selected: boolean)     |         |
-| onItemSelectAll | Select a group of items | (keys: string\[], selected: boolean) |         |
+| Property | Description | Type | Version |
+| --- | --- | --- | --- |
+| direction | List render direction | `left` \| `right` |  |
+| disabled | Disable list or not | boolean |  |
+| filteredItems | Filtered items | RecordType\[] |  |
+| selectedKeys | Selected items | string\[] |  |
+| onItemSelect | Select item | (key: string, selected: boolean, event?: React.MouseEvent) |  |
+| onItemSelectAll | Select a group of items | (keys: string\[], selected: boolean) |  |
```
Comment (yoyo837) [inline components/transfer/index.en-US.md:78]: 文档是否需要说明一下shift操作?

Comment (zbw-zbw) [reply components/transfer/index.en-US.md:78]: > 文档是否需要说明一下shift操作?

嗯嗯 我加下

Comment (zbw-zbw) [reply components/transfer/index.en-US.md:78]: done!

File: components/table/hooks/useSelection.tsx [hunk 2]
```diff
@@ -350,7 +351,7 @@ const useSelection = <RecordType extends AnyObject = AnyObject>(
         ...selection,
         onSelect: (...rest) => {
           selection.onSelect?.(...rest);
-          setLastSelectedKey(null);
+          updatePrevSelectedIndex(-1);
```
Comment (zombieJ) [inline components/table/hooks/useSelection.tsx:354]: 这边还是推荐用 `null`，`-1` 散在代码中很容易被重构搞坏掉

File: components/_util/hooks/useMultipleSelect.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * @title multipleSelect hooks
+ * @description multipleSelect by hold down shift key
+ */
+export default function useMultipleSelect() {
+  const prevSelectedIndexRef = useRef(-1);
+
+  const multipleSelect = <T>(currentSelectedIndex: number, data: any[], selectedKeys: Set<T>) => {
+    // prevSelectedIndex reset case
+    if (prevSelectedIndexRef.current < 0) {
```
[the remark this answers is not in the document]
Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts:12]: done

File: components/_util/hooks/useMultipleSelect.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  const multipleSelect = useCallback(
+    <T>(currentSelectedIndex: number, data: any[], selectedKeys: Set<T>) => {
+      let configPrevSelectedIndex = prevSelectedIndex;
+
+      // prevSelectedIndex reset case
+      if (prevSelectedIndex === null) {
+        const selectedIndexArr: number[] = [];
+        data.forEach((item, idx) => {
+          if (Array.from(selectedKeys).includes(item?.key ?? item)) {
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:20]: 这个有性能问题，应该是提前生成一个 Set 然后判断 has 做命中

File: components/_util/hooks/useMultipleSelect.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * @title multipleSelect hooks
+ * @description multipleSelect by hold down shift key
+ */
+export default function useMultipleSelect() {
+  const [prevSelectedIndex, setPrevSelectedIndex] = useState<PrevSelectedIndex>(null);
+
+  const multipleSelect = useCallback(
+    <T>(currentSelectedIndex: number, data: any[], selectedKeys: Set<T>) => {
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:13]: `any` 类型不靠谱，要抽成泛型放 `useMultipleSelect` 上

Comment (zombieJ) [reply components/_util/hooks/useMultipleSelect.ts:13]: `selectedKeys` 类型理论上也是在 `useMultipleSelect` 就可以定义的

File: components/_util/hooks/useMultipleSelect.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      if (prevSelectedIndex === null) {
+        const selectedIndexArr: number[] = [];
+        data.forEach((item, idx) => {
+          if (Array.from(selectedKeys).includes(item?.key ?? item)) {
+            selectedIndexArr.push(idx);
+          }
+        });
+
+        // nearest item between currentIndex and selectedIndexArr
+        let nearestIndex = selectedIndexArr[0];
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:26]: 为啥会需要一个 nearestIndex?

File: components/_util/hooks/useMultipleSelect.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            nearestIndex = item;
+          }
+        });
+        configPrevSelectedIndex = nearestIndex ?? currentSelectedIndex;
+      }
+
+      // add/delete the selected range
+      const startIndex = Math.min(configPrevSelectedIndex || 0, currentSelectedIndex);
+      const endIndex = Math.max(configPrevSelectedIndex || 0, currentSelectedIndex);
+      const rangeKeys = data.slice(startIndex, endIndex + 1).map((item) => item?.key ?? item);
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:40]: `item?.key ?? item` 这个写法很诡异，感觉是为了兼容两个凑出来的。应该是 hook 提供一个 getKey 方法来做判断

File: components/_util/hooks/useMultipleSelect.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      const shouldSelected = rangeKeys.some((key) => !selectedKeys?.has(key));
+      const changedKeys: T[] = [];
+
+      rangeKeys.forEach((item) => {
+        if (shouldSelected) {
+          if (!selectedKeys.has(item)) {
+            changedKeys.push(item);
+          }
+          selectedKeys.add(item);
+          setPrevSelectedIndex(endIndex);
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:50]: 不应该在 forEach 下反复调用 setState

File: components/_util/hooks/useMultipleSelect.ts (hunk 6 above)
Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts:26]: nearestIndex 是为了 prevSelectedIndex 为空时能够继续多选的操作，找到和当前选中位置最近的一个已选中位置作为一个 range

File: components/_util/hooks/useMultipleSelect.ts (hunk 5 above)
Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: > `selectedKeys` 类型理论上也是在 `useMultipleSelect` 就可以定义的

这样子的话是要在初始化调用 hooks 时就要传参了吧 好像不太行耶

File: components/_util/hooks/useMultipleSelect.ts (hunk 6 above)
Comment (zombieJ) [reply components/_util/hooks/useMultipleSelect.ts:34]: 我理解中，如果 prev 是空的。那应该就不要执行范围选择的逻辑才对。没必要自动选一个，因为用户是不能感知这个逻辑的。

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts:34]: 嗯嗯 我懂你的意思。之前做这个逻辑是考虑到这种场景，方便用户多选：

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts:34]: 上面提到的这种场景 用户是不是就可以感知到这个逻辑了

Comment (zombieJ) [reply components/_util/hooks/useMultipleSelect.ts:34]: shift 操作这个很难讲，因为它可能是要全部取消。你如果给了范围，它取反即可。但是没有的话这个是说不准的。

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts:34]: 行 那我待会把这部分逻辑干掉

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: done

File: components/_util/hooks/useMultipleSelect.ts (hunk 8 above)
Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: done

File: components/_util/hooks/useMultipleSelect.ts (hunk 7 above)
Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: done

File: components/_util/hooks/useMultipleSelect.ts (hunk 4 above)
Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: done

File: components/_util/hooks/useMultipleSelect.ts (hunk 5 above)
Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: done

File: components/_util/hooks/useMultipleSelect.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * @title multipleSelect hooks
+ * @description multipleSelect by hold down shift key
+ */
+export default function useMultipleSelect() {
+  const [prevSelectedIndex, setPrevSelectedIndex] = useState<PrevSelectedIndex>(null);
+
+  const multipleSelect = useCallback(
+    <T, K>(currentSelectedIndex: number, data: T[], selectedKeys: Set<K>, key?: (item: T) => K) => {
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:13]: 这个泛型能挪上去不，useMultipleSelec<...>。`multipleSelect` 上加泛型语义上很奇怪，每个调用的地方都可以自己搞一套不同的类型。实际上在使用 hooks 的时候就已经定下了

Comment (zombieJ) [reply components/_util/hooks/useMultipleSelect.ts:13]: `key` 这个方法位置不对，应该也是 `useMultipleSelect` 的参数，而不是 `multipleSelect` 的参数

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: done

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: > `key` 这个方法位置不对，应该也是 `useMultipleSelect` 的参数，而不是 `multipleSelect` 的参数

好 我再处理下

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: > `key` 这个方法位置不对，应该也是 `useMultipleSelect` 的参数，而不是 `multipleSelect` 的参数

done

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: > `key` 这个方法位置不对，应该也是 `useMultipleSelect` 的参数，而不是 `multipleSelect` 的参数

done

File: components/_util/hooks/useMultipleSelect.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+export type PrevSelectedIndex = null | number;
+
+/**
+ * @title multipleSelect hooks
+ * @description multipleSelect by hold down shift key
+ */
+export default function useMultipleSelect<T, K>(key?: (item: T) => K) {
+  const [prevSelectedIndex, setPrevSelectedIndex] = useState<PrevSelectedIndex>(null);
+
+  const getKey = (item: T): K => {
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:12]: 这个需要 hooks 调用这一定要提供 `getKey` 方法，下面不能 `return item as unknown as K;`

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts:12]: 行 这里之前是想着默认拿 item 作为 key 那我把key改成必传参数吧 比较合理

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: done!

File: components/_util/hooks/useMultipleSelect.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  const [prevSelectedIndex, setPrevSelectedIndex] = useState<PrevSelectedIndex>(null);
+
+  const multipleSelect = useCallback(
+    (currentSelectedIndex: number, data: T[], selectedKeys: Set<K>) => {
+      const configPrevSelectedIndex = prevSelectedIndex ?? currentSelectedIndex;
+
+      // add/delete the selected range
+      const startIndex = Math.min(configPrevSelectedIndex || 0, currentSelectedIndex);
+      const endIndex = Math.max(configPrevSelectedIndex || 0, currentSelectedIndex);
+      const rangeKeys = data.slice(startIndex, endIndex + 1).map((item) => getKey?.(item));
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:19]: 上面已经保证提供，这里就不用 `?.`  了

File: components/_util/hooks/useMultipleSelect.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+      const shouldSelected = rangeKeys.some((rangeKey) => !selectedKeys?.has(rangeKey));
```
Comment (zombieJ) [inline components/_util/hooks/useMultipleSelect.ts:20]: 这里 `selectedKeys` 在定义的时候也已经保证提供了，不需要 `?.`。其他的地方也检查一下有没有可以删掉的~

Comment (zbw-zbw) [reply components/_util/hooks/useMultipleSelect.ts]: 好 看了下应该就这俩地方了 done！

File: components/modal/demo/dark.tsx [hunk 13]
```diff
@@ -258,8 +258,8 @@ const TableTransfer = ({ leftColumns, rightColumns, ...restProps }) => (
             : difference(listSelectedKeys, treeSelectedKeys);
           onItemSelectAll(diffKeys, selected);
         },
-        onSelect({ key }, selected) {
-          onItemSelect(key, selected);
+        onSelect({ key }, selected, _selectedRows, nativeEvent) {
+          onItemSelect(key, selected, nativeEvent);
```
Comment (zombieJ) [inline components/modal/demo/dark.tsx:262]: 这个 demo 没有清理

File: components/transfer/ListBody.tsx [hunk 14]
```diff
@@ -73,8 +73,8 @@ const TransferListBody: React.ForwardRefRenderFunction<
     }
   }, [filteredRenderItems, mergedPagination, pageSize]);
 
-  const onClick = (item: RecordType) => {
-    onItemSelect?.(item.key, !selectedKeys.includes(item.key));
+  const handleClick = (item: RecordType, e: React.MouseEvent<Element, MouseEvent>) => {
```
Comment (zombieJ) [inline components/transfer/ListBody.tsx:76]: 不要用 handle，可以叫 `onInternalClick`

File: components/transfer/list.tsx [hunk 15]
```diff
@@ -52,6 +53,7 @@ export interface TransferListProps<RecordType> extends TransferLocale {
   style?: React.CSSProperties;
   checkedKeys: string[];
   handleFilter: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  onClick: (key: string, check: boolean, e: React.MouseEvent<Element, MouseEvent>) => void;
```
Comment (zombieJ) [inline components/transfer/list.tsx:56]: 这边不应该加 onClick 事件，而是应该在 `renderListBody` 里的 `renderList(listProps)` 对 onSelectItem 进行清理

Comment (zombieJ) [reply components/transfer/list.tsx:56]: 否则这个仍然是把内部的方法暴露出去了

File: components/transfer/list.tsx [hunk 16]
```diff
@@ -133,11 +138,14 @@ const TransferList = <RecordType extends KeyWiseTransferItem>(
     return text.includes(filterValue);
   };
 
-  const renderListBody = (listProps: TransferListBodyProps<RecordType>) => {
+  const renderListBody = (defaultListProps: TransferListBodyProps<RecordType>) => {
+    const { onItemSelect: onInternalItemSelect, ...restProps } = defaultListProps;
+    const onItemSelect = (key: string, check: boolean) => onInternalItemSelect(key, check);
+    const listProps: TransferCustomListBodyProps<RecordType> = { ...restProps, onItemSelect };
```
Comment (zombieJ) [inline components/transfer/list.tsx:144]: 合并一下：
```tsx
let bodyContent: React.ReactNode = renderList ? renderList({
   ...listProps,
   onItemSelect: (key: string, check: boolean) => defaultListProps.onItemSelect(key: string, check: boolean),
}) : null;
```

Comment (zombieJ) [reply components/transfer/list.tsx:144]: 这里也不用叫 defaultListProps 了，还原会 listProps 即可

**Review by zombieJ — APPROVED** [review]

**Pull request merged**